rpc/rpc.go

```diff
@@ -138,82 +132,104 @@ type ErrorReporter interface {
 // requests from the transport.
 func NewConn(t Transport, opts *Options) *Conn {
 	bgctx, bgcancel := context.WithCancel(context.Background())
+
+	// We use an errgroup to link the lifetime of background tasks
+	// to each other.
+	g, bgctx := errgroup.WithContext(bgctx)
+
 	c := &Conn{
 		transport: t,
 		shut:      make(chan struct{}),
 		bgctx:     bgctx,
 		bgcancel:  bgcancel,
 		answers:   make(map[answerID]*answer),
 		imports:   make(map[importID]*impent),
+		sendq:     mpsc.New(),
 	}
 	if opts != nil {
 		c.bootstrap = opts.BootstrapClient
-		c.reporter = opts.ErrorReporter
+		c.er.ErrorReporter = opts.ErrorReporter
 		c.abortTimeout = opts.AbortTimeout
 	}
 	if c.abortTimeout == 0 {
 		c.abortTimeout = 100 * time.Millisecond
 	}
-	c.tasks.Add(1)
+
+	// start background tasks
+	g.Go(c.newTask(c.send))
+	g.Go(c.newTask(c.receive))
+
+	// monitor background tasks
 	go func() {
-		abortErr := c.receive(c.bgctx)
-		c.tasks.Done()
+		if err := g.Wait(); err != nil {
+			c.mu.Lock() // shutdown unlocks c.mu.
 
-		c.mu.Lock()
-		select {
-		case <-c.bgctx.Done():
-			c.mu.Unlock()
-		default:
-			if abortErr != nil {
-				c.report(abortErr)
-			}
-			// shutdown unlocks c.mu.
-			if err := c.shutdown(abortErr); err != nil {
-				c.report(err)
+			c.er.ReportError(err)
+			if err = c.shutdown(err); err != nil {
+				c.er.ReportError(err)
 			}
 		}
 	}()
+
 	return c
 }
 
+func (c *Conn) newTask(f func(context.Context) error) func() error {
+	return func() error {
+		defer c.tasks.Done()
```

Comment:
Where is the corresponding call to c.tasks.Add()? Maybe this should just be deleted now that newTask is called inside g.Go? But I don't think that's compatible with using startTask() elsewhere, so maybe not.